### PR TITLE
Rewrite of MNIST example

### PR DESCRIPTION
**Pull request by f0k**
Includes both the dense and the convolutional architecture in a single example,
reduces indirections in the code and adds a lot of explanations (addressing #215, #322).

Also changes the code to not copy the dataset to GPU at once, but in batches -- this is easier to understand and modify for large datasets or on-the-fly augmentation. While at it, it also changes the code to shuffle the training data at each epoch. Finally, it changes the batch size from 600 to 500 so it evenly divides the train/val/test sizes (50,000, 10,000 and 10,000 examples each).

Although each of those changes should negatively affect performance, performance is almost the same as before (note: you need to run it as `mnist.py custom_mlp:2,512,0,0.5` to get the same architecture as in the original script).

The only question is whether we want the example file to be as extensively commented as I did now, or better reduce comments to a minimum and move the text to http://lasagne.readthedocs.org/en/latest/user/tutorial.html, interspersed with code snippets. I'm tending towards the latter, but I'm not sure if the example file still remains a good read then.

### CODE REVIEW

File: examples/mnist.py [hunk 1]
```diff
@@ -1,240 +1,414 @@
...
+        print("       'custom_mlp:DEPTH,WIDTH,DROP_IN,DROP_HID' for an MLP")
+        print("       with DEPTH hidden layers of WIDTH units, DROP_IN")
+        print("       input dropout and DROP_HID hidden dropout,")
+        print("       'cnn' for a simple Convolutional Neural Network (CNN).")
+        print("EPOCHS: number of training epochs to perform (default: 500)")
+        return
+    if len(sys.argv) > 1:
+        model = sys.argv[1]
+    if len(sys.argv) > 2:
+        num_epochs = int(sys.argv[2])
```
Comment (dnouri) [inline examples/mnist.py]: I guess let's move checking of `sys.argv` to the `if __name__ == '__main__'` block below.

Comment (f0k) [inline examples/mnist.py]: Sounds good. Changed that (and included the change to `test_examples.py` I forgot to commit).

**Comment by benanne** [comment]
Great work! I'll try to have an in-depth look at this ASAP.

**Comment by f0k** [comment]
> I think I agree with your idea of moving most of the prose into a tutorial section. It also looks better if the example code can be 'glanced' :)

Did that now. It replicates a bit of what's in `docs/layers.rst`, but I think it's practical to explain right on spot how layers work. To distinguish the two, `docs/layers.rst` could be expanded a bit to include DAGs (i.e., branching by stacking multiple layers on top of the same source, and merging via one of the merge layers). That's not exactly critical for the first release, though.

Would be nice if somebody could at least superficially review my example and the tutorial that goes along with it.

**Comment by moscow25** [comment]
As someone who recently (6 months back) got a running start from the MNIST
example, I both liked it as it were, and would have been even happier with
more inline commentary that I'm forced to see. It's not obvious at all, to
a beginner, what the different networks are than can be setup very quickly.
(Similarly, it wasn't clear to me that I don't need to load all my data
into theano.shared(), etc.)

Not sure it's possible to "glance" the code, in any case, and explaining
the options really does help provide a reference. The further that the
documentation goes from inline to out-of-line, the less likely it is to be
read, or maintained. I love the long explanations in "layers," "updates,"
etc.

Just my $0.02.

Great job, and thanks as always.

Best,
Nikolai

**Comment by benanne** [comment]
I should have clarified that ASAP might not have been very soon at all... sorry! I was at ICML for the past week and I'm moving abroad this week so I can't help much for now. Anyone else up for it?

**Comment by f0k** [comment]
> As someone who recently (6 months back) got a running start from the MNIST example, I both liked it as it were, and would have been even happier with more inline commentary that I'm forced to see.

Hmm, interesting! I thought it was really a bit hard to read. The way it is now the example code should be understandable to anybody familiar with neural networks and Theano, and the tutorial text gives more in-depth explanations and pointers where to learn about the preliminaries if needed.

> It's not obvious at all, to a beginner, what the different networks are than can be setup very quickly.

That's right, but teaching neural networks is clearly outside the scope of the example. We might add some neural network tutorials to [Lasagne/Recipes](https://github.com/Lasagne/Recipes), but this example targets people that know about neural networks and Theano and mostly want to learn about Lasagne.

> explaining the options really does help provide a reference

With moving most explanations from the code into the Sphinx docs, it's much easier to reference the Lasagne and Theano docs (e.g., the `InputLayer` or `DenseLayer` docstring, or what a Theano variable is).

> The further that the documentation goes from inline to out-of-line, the less likely it is to be read,

That's right, although the example provides a link to the tutorial, and the README file will probably also link to it. I could try to make the links more prominent.

> or maintained.

Yes, maintenance is clearly a pain in that form, because some of the code is now duplicated between the Python script and the ReST file. But the hope is that this will hardly need any changes, just like Theano's deep learning tutorials.

Thanks a lot for your input!

> Anyone else up for it?

Maybe @craffel, since he also wrote a tutorial at some point? (One that I really like, by the way, it would just need some updates to comply with API changes.) Do you want to have a quick look at the example and explanations? It doesn't need to be perfect for now, it just needs to be better than what we have.

File: docs/user/tutorial.rst [hunk 2]
```diff
@@ -4,88 +4,611 @@
...
+``y_train.shape`` is simply ``(50000,)``, that is, it is a vector the same
+length of ``X_train`` giving an integer class label for each image -- namely,
+the digit between 0 and 9 depicted in the image (according to the human
+annotator who drew that digit).
+
+
+Building the model
+------------------
+
+This is were Lasagne steps in. It allows you to define an arbitrarily
```
Comment (ebenolson) [inline docs/user/tutorial.rst]: s/were/where/

File: docs/user/tutorial.rst [hunk 3]
```diff
@@ -4,88 +4,611 @@
...
+
+    def build_mlp(input_var=None):
+        l_in = lasagne.layers.InputLayer(shape=(None, 1, 28, 28),
+                                         input_var=input_var)
+
+The four numbers in the shape tuple represent, in order:
+``(batchsize, channels, rows, columns)``.
+Here we've set the batchsize to ``None``, which means the network will accept
+input data of arbitrary batchsize after compilation. If you know the batchsize
+beforehands and do not need this flexibility, you should give the batchsize
```
Comment (ebenolson) [inline docs/user/tutorial.rst]: s/beforehands/beforehand/

File: docs/user/tutorial.rst [hunk 4]
```diff
@@ -4,88 +4,611 @@
...
+The function begins like the others:
+
+.. code-block:: python
+
+    def build_cnn(input_var=None):
+        layer = lasagne.layers.InputLayer(shape=(None, 1, 28, 28),
+                                          input_var=input_var)
+
+We don't apply dropout to the inputs, as this tends to work less well for
+convolutional layers. Besides, instead of a :class:`DenseLayer
```
Comment (ebenolson) [inline docs/user/tutorial.rst]: I'd remove 'Besides' here. Also, might be a little confusing using both 'kernels' and 'filters'.

File: docs/user/tutorial.rst [hunk 5]
```diff
@@ -4,88 +4,611 @@
...
+Equipped with all the necessary Theano expressions, we're now ready to compile
+a function performing a training step:
+
+.. code-block:: python
+
+    train_fn = theano.function([input_var, target_var], loss, updates=updates)
+
+This tells Theano to generate and compile a function taking two inputs -- a
+mini-batch of images and a vector of corresponding targets -- and returning a
+single output: the training loss. Besides, each time it is invoked, it applies
```
Comment (ebenolson) [inline docs/user/tutorial.rst]: I think think would read better with 'Additionally' instead of 'Besides'.

**Comment by ebenolson** [comment]
:+1:  from me, for a basic example I think this is a big improvement. 

Aesthetically, my preference would be an IPython notebook, which could then be automatically converted to .py and [rst](http://sphinx-ipynb.readthedocs.org/en/latest/hello.html), but I think this arrangement is fine.

**Comment by f0k** [comment]
Thanks for reviewing! I've addressed all your comments.

> Aesthetically, my preference would be an IPython notebook, which could then be automatically converted to .py and rst

Sounds good in theory, but that would give a little less flexibility -- now we can freely select and shorten the code excerpts to use in the tutorial, similar to Theano's deep learning tutorial. An IPython notebook might still be a good alternative to the Sphinx doc. It could do everything in a flat file, without any function definitions (except maybe for `iterate_minibatches()`), and just for the MLP. `mnist.py` would then be a somewhat more fleshed-out variant of what the notebook does, with three networks and command line support. I also like Colin's tutorial with its references to cooking. Countless possibilities!

I'll leave this open until tomorrow and then squash and merge it if there are no further comments.

**Comment by craffel** [comment]
> Maybe @craffel, since he also wrote a tutorial at some point? (One that I really like, by the way, it would just need some updates to comply with API changes.)

It's up to date as of about two weeks ago, and I can update it again if needed.  I've given it as a tutorial 3 times and people seem to like it.  I have a version which is not online yet which includes recurrent layers too.

http://nbviewer.ipython.org/github/craffel/Lasagne-tutorial/blob/master/examples/tutorial.ipynb

**Comment by f0k** [comment]
> It's up to date as of about two weeks ago, and I can update it again if needed.

But the version you linked to still uses `Layer.get_output()`, and the `Objective` class. You should try running it without `warnings.filterwarnings` once to see which interface changes need to be incorporated to be ready for the Lasagne release.

**Comment by craffel** [comment]
> But the version you linked to still uses Layer.get_output(), and the Objective class. You should try running it without warnings.filterwarnings once to see which interface changes need to be incorporated to be ready for the Lasagne release.

Oh, sure, right.  I can/will change those.
